Lib/pathlib.py

```diff
@@ -1385,6 +1385,132 @@ def expanduser(self):
+        However, the caller cannot modify the child_directory_names to prune
+        the search or to impose a specific order of visiting because the list
+        of directories to visit is calculated before we yield it.
+        """
+        sys.audit("pathlib.Path.walk_bottom_up", self, on_error, follow_links)
+        return self._walk(False, on_error, follow_links)
+
+    def _walk(self, topdown, on_error, follow_links):
+        dirs = []
+        nondirs = []
+        walk_dirs = []
+
+        # We may not have read permission for self, in which case we can't
+        # get a list of the files the directory contains. os.walk
+        # always suppressed the exception then, rather than blow up for a
+        # minor reason when (say) a thousand readable directories are still
+        # left to visit. That logic is copied here.
+        try:
+            scandir_it = self._scandir()
+        except OSError as error:
+            if on_error is not None:
+                on_error(error)
+            return
+
+        with scandir_it:
+            while True:
+                try:
+                    entry = next(scandir_it, ...)
+                    if entry is ...:
+                        break
+                except OSError as error:
+                    if on_error is not None:
+                        on_error(error)
+                    return
+
+                try:
+                    is_dir = entry.is_dir()
+                except OSError:
+                    # If is_dir() raises an OSError, consider that the entry
+                    # is not a directory, same behavior as os.path.isdir()
+                    is_dir = False
+
+                if is_dir:
+                    dirs.append(entry.name)
+                else:
+                    nondirs.append(entry.name)
+
+                if not topdown and is_dir:
+                    # Bottom-up: recurse into sub-directory, but exclude symlinks to
+                    # directories if follow_links is False
+                    if follow_links:
+                        walk_into = True
+                    else:
+                        try:
+                            is_symlink = entry.is_symlink()
+                        except OSError:
+                            # If is_symlink() raises an OSError, consider that
+                            # the entry is not a symbolic link, same behavior
+                            # as os.path.islink()
+                            is_symlink = False
+                        walk_into = not is_symlink
+
+                    if walk_into:
+                        walk_dirs.append(entry)
+        if topdown:
+            for raw_new_path in dirs:
+                # Path.is_symlink() is used instead of caching entry.is_symlink()
+                # result during the loop on Path._scandir() because the caller
+                # can replace the directory entry during the "yield" above.
```

Comment:
If the caller adds or removes a directory during the walk, its their responsibility to update `dirnames` IMO. I don't think it's worth an extra `is_symlink()` here.

Reply:
True. Though implementation loses a feature and becomes slightly more complicated. The benefit of this is that top down walk is now actually slightly faster than os.walk. But is it worth it? I would like to hear people's use cases. Maybe replacement with a symlink without check and in follow_symlinks=False is a lot more common use case than you and I think.

Reply:
It becomes *less* complicated, not more. The end of the method becomes:

```python
        if top_down:
            yield self, dirnames, filenames
        for dirname in dirnames:
            yield from self._make_child_relpath(dirname).walk(
                top_down=top_down, follow_symlinks=follow_symlinks, on_error=on_error)
        if not top_down:
            yield self, dirnames, filenames
```

Reply:
Yes, you are right. Not sure how I didn't notice that